pyteal/ast/abi/util.py

```diff
@@ -499,3 +499,47 @@ def type_specs_from_signature(sig: str) -> tuple[list[TypeSpec], Optional[TypeSp
         return_type = type_spec_from_algosdk(sdk_method.returns.type)
 
     return [type_spec_from_algosdk(arg.type) for arg in sdk_method.args], return_type
+
+
+def type_spec_is_assignable(a: TypeSpec, b: TypeSpec) -> bool:
+    from pyteal.ast.abi import (
+        TupleTypeSpec,
+        ArrayTypeSpec,
+        StaticArrayTypeSpec,
+        DynamicArrayTypeSpec,
+    )
+
+    match a, b:
+        case TupleTypeSpec(), TupleTypeSpec():
+            a, b = cast(TupleTypeSpec, a), cast(TupleTypeSpec, b)
+            if len(a.value_type_specs()) != len(b.value_type_specs()):
+                return False
+            return all(
+                map(
+                    lambda ab: type_spec_is_assignable(ab[0], ab[1]),
+                    zip(a.value_type_specs(), b.value_type_specs()),
+                )
+            )
+        case TupleTypeSpec(), _:
+            return False
+        case ArrayTypeSpec(), ArrayTypeSpec():
+            a, b = cast(ArrayTypeSpec, a), cast(ArrayTypeSpec, b)
+            if not type_spec_is_assignable(a.value_type_spec(), b.value_type_spec()):
+                return False
+            match a, b:
+                case StaticArrayTypeSpec(), StaticArrayTypeSpec():
+                    a, b = cast(StaticArrayTypeSpec, a), cast(StaticArrayTypeSpec, b)
+                    return a.length_static() == b.length_static()
+                case DynamicArrayTypeSpec(), DynamicArrayTypeSpec():
+                    return True
+                case _:
+                    return False
+        case ArrayTypeSpec(), _:
+            return False
+
+    if isinstance(a, type(b)):
```

Comment:
This case is taken care of in latest commit.

Reply:
It would be captured in earlier `match`, and should not get into this position of code.

Reply:
Thanks for clarifying. I just want to point out that this is a potential failure point if we were to introduce new container types.

Reply:
Open to discussion, but want to leave the note because some time passed + we've had at least one more verbal.

Assuming there's no obvious + easy type check change, I still think it's relatively moderate cost to implement a check like https://github.com/algorand/pyteal/blob/fdf10f7c7dad4741ceec84b07413211a84086210/pyteal/ast/abi/util_test.py#L395-L404

* Since returning `true` when _not_ assignable is likely the more dangerous situation, I can imagine strengthening the analogous check by requiring the expected result to be `false` (unassignable). 
  * In the parlance of https://github.com/algorand/pyteal/pull/548/, it means looking for at least one `UnsafeAssignment` or `UnsafeBidirectional`.
* The proposal feels easier to maintainer from my perspective than a more automated check that attempts to traverse the type system hierarchy.  Obviously, the proposal leaves room for error, but I've not been able to convince myself that a better tradeoff exists.